HBASE-23693 Split failure may cause region hole and data loss when use zk assign

thangTang opened the pull request:
https://issues.apache.org/jira/browse/HBASE-23693

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 28s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -0 :warning: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ branch-1 Compile Tests _ |
| +0 :ok: |  mvndep  |   1m 22s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   7m 41s |  branch-1 passed  |
| +1 :green_heart: |  compile  |   1m  7s |  branch-1 passed with JDK v1.8.0_242  |
| +1 :green_heart: |  compile  |   1m 11s |  branch-1 passed with JDK v1.7.0_252  |
| +1 :green_heart: |  checkstyle  |   2m 27s |  branch-1 passed  |
| +1 :green_heart: |  shadedjars  |   3m 15s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 55s |  branch-1 passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javadoc  |   1m  9s |  branch-1 passed with JDK v1.7.0_252  |
| +0 :ok: |  spotbugs  |   2m 54s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   4m 24s |  branch-1 passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 15s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m  6s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  1s |  the patch passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javac  |   1m  1s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 11s |  the patch passed with JDK v1.7.0_252  |
| +1 :green_heart: |  javac  |   1m 11s |  the patch passed  |
| -1 :x: |  checkstyle  |   1m 43s |  hbase-server: The patch generated 1 new + 18 unchanged - 0 fixed = 19 total (was 18)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   3m  6s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |   5m 21s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2.  |
| +1 :green_heart: |  javadoc  |   0m 51s |  the patch passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javadoc  |   1m 10s |  the patch passed with JDK v1.7.0_252  |
| -1 :x: |  findbugs  |   1m 40s |  hbase-client generated 2 new + 0 unchanged - 0 fixed = 2 total (was 0)  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   2m 38s |  hbase-client in the patch passed.  |
| -1 :x: |  unit  | 151m 52s |  hbase-server in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 46s |  The patch does not generate ASF License warnings.  |
|  |   | 206m 37s |   |


| Reason | Tests |
|-------:|:------|
| FindBugs | module:hbase-client |
|  |  Dead store to t in org.apache.hadoop.hbase.MetaTableAccessor.getDaughterRegionsFromParent(Connection, HRegionInfo)  At MetaTableAccessor.java:org.apache.hadoop.hbase.MetaTableAccessor.getDaughterRegionsFromParent(Connection, HRegionInfo)  At MetaTableAccessor.java:[line 852] |
|  |  Dead store to t in org.apache.hadoop.hbase.MetaTableAccessor.getHRegionInfo(Connection, byte[])  At MetaTableAccessor.java:org.apache.hadoop.hbase.MetaTableAccessor.getHRegionInfo(Connection, byte[])  At MetaTableAccessor.java:[line 845] |
| Failed junit tests | hadoop.hbase.replication.TestReplicationKillSlaveRS |
|   | hadoop.hbase.master.TestMasterOperationsForRegionReplicas |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1071/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1071 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux baa1866f84de 4.15.0-74-generic #84-Ubuntu SMP Thu Dec 19 08:06:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/Base-PreCommit-GitHub-PR_PR-1071/out/precommit/personality/provided.sh |
| git revision | branch-1 / 64b9d2f |
| Default Java | 1.7.0_252 |
| Multi-JDK versions | /usr/lib/jvm/zulu-8-amd64:1.8.0_242 /usr/lib/jvm/zulu-7-amd64:1.7.0_252 |
| checkstyle | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1071/1/artifact/out/diff-checkstyle-hbase-server.txt |
| findbugs | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1071/1/artifact/out/new-findbugs-hbase-client.html |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1071/1/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1071/1/testReport/ |
| Max. process+thread count | 3950 (vs. ulimit of 10000) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1071/1/console |
| versions | git=1.9.1 maven=3.0.5 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 28s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -0 :warning: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ branch-1 Compile Tests _ |
| +0 :ok: |  mvndep  |   1m 21s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   7m 33s |  branch-1 passed  |
| +1 :green_heart: |  compile  |   1m  1s |  branch-1 passed with JDK v1.8.0_242  |
| +1 :green_heart: |  compile  |   1m 12s |  branch-1 passed with JDK v1.7.0_252  |
| +1 :green_heart: |  checkstyle  |   2m 24s |  branch-1 passed  |
| +1 :green_heart: |  shadedjars  |   3m 14s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 58s |  branch-1 passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javadoc  |   1m  8s |  branch-1 passed with JDK v1.7.0_252  |
| +0 :ok: |  spotbugs  |   2m 57s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   4m 29s |  branch-1 passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 15s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m 10s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  0s |  the patch passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javac  |   1m  0s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 11s |  the patch passed with JDK v1.7.0_252  |
| +1 :green_heart: |  javac  |   1m 11s |  the patch passed  |
| -1 :x: |  checkstyle  |   0m 38s |  hbase-client: The patch generated 1 new + 79 unchanged - 0 fixed = 80 total (was 79)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   3m  9s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |   5m 16s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2.  |
| +1 :green_heart: |  javadoc  |   0m 51s |  the patch passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javadoc  |   1m  8s |  the patch passed with JDK v1.7.0_252  |
| +1 :green_heart: |  findbugs  |   4m 43s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   2m 37s |  hbase-client in the patch passed.  |
| +1 :green_heart: |  unit  | 152m 18s |  hbase-server in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 45s |  The patch does not generate ASF License warnings.  |
|  |   | 206m 25s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1071/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1071 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 514f8208f1d5 4.15.0-74-generic #84-Ubuntu SMP Thu Dec 19 08:06:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/Base-PreCommit-GitHub-PR_PR-1071/out/precommit/personality/provided.sh |
| git revision | branch-1 / 99a59cf |
| Default Java | 1.7.0_252 |
| Multi-JDK versions | /usr/lib/jvm/zulu-8-amd64:1.8.0_242 /usr/lib/jvm/zulu-7-amd64:1.7.0_252 |
| checkstyle | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1071/2/artifact/out/diff-checkstyle-hbase-client.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1071/2/testReport/ |
| Max. process+thread count | 3931 (vs. ulimit of 10000) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1071/2/console |
| versions | git=1.9.1 maven=3.0.5 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

wchevreuil commented on hbase-server/src/main/java/org/apache/hadoop/hbase/master/RegionStates.java:
```diff
@@ -783,10 +801,20 @@ public void regionOffline(
+              HRegionInfo parent = daughter2Parent.get(hri.getEncodedName());
+              HRegionInfo info = getHRIFromMeta(parent);
+              if (info != null && info.isSplit() && info.isOffline()) {
+                regionsToClean.add(Pair.newPair(state.getRegion(), info));
```
Oh, so parent is **OFFLINE** and **SPLIT=true** in meta only, but not in _assignedRegions_? Sounds wrong that we marked parent as OFFLINE and SPLIT even when splitting was still not entirely complete.

thangTang replied:
If the RS machine crash when the SplitTransactionImpl step after PONR, master will handle the split rollback. Under normal conditions, it cleanup daughter region and try to assign parent region. If assign failed, parent region RIT so CatalogJanitor blocked; but at this time the master switch, new master will delete parent region because of **OFFLINE and SPLIT=true**.

saintstack replied:
This daughter2Parent added stingency makes sense to me. Nice.

wchevreuil replied:
> If the RS machine crash when the SplitTransactionImpl step after PONR, master will handle the split rollback.

Yeah, but it seems to work only because Master has different info for the region than what is actual in meta. What if Active Master also crashes before it triggers SCP and daughters get cleaned up? The other master will load regions state as seen in meta, then it can run into same problem again.

thangTang replied:
> > If the RS machine crash when the SplitTransactionImpl step after PONR, master will handle the split rollback.
> 
> Yeah, but it seems to work only because Master has different info for the region than what is actual in meta. What if Active Master also crashes before it triggers SCP and daughters get cleaned up? The other master will load regions state as seen in meta, then it can run into same problem again.

daughter cleanup is excuted in SCP, and in this patch it will do three things together:
1.delete daughter region in meta
2.update parent region in meta
3.delete daughter region dir in HDFS
`	          if (regionPair != null) {
            MetaTableAccessor.deleteRegion(this.server.getConnection(), hri);	            MetaTableAccessor.deleteRegion(this.server.getConnection(), hri);
          }	          }
          if (parentInfo != null) {
            List<Mutation> mutations = new ArrayList<Mutation>();
            HRegionInfo copyOfParent = new HRegionInfo(parentInfo);
            copyOfParent.setOffline(false);
            copyOfParent.setSplit(false);
            Put putParent = MetaTableAccessor.makePutFromRegionInfo(copyOfParent);
            mutations.add(putParent);
            MetaTableAccessor.mutateMetaTable(this.server.getConnection(), mutations);
          }
          LOG.debug("Cleaning up HDFS since no meta entry exists, hri: " + hri);	          LOG.debug("Cleaning up HDFS since no meta entry exists, hri: " + hri);
          FSUtils.deleteRegionDir(server.getConfiguration(), hri);	          FSUtils.deleteRegionDir(server.getConfiguration(), hri);`

So if Active Master also crashes before it triggers SCP, the daughter won`t be deleted.

wchevreuil replied:
> So if Active Master also crashes before it triggers SCP, the daughter won`t be deleted.

Yes, that's my point. We would now have a new active master that sees parent split as complete, although split was midway through. It will potentially remove the parent, and fail to online the daughters. 

I wonder if working on correcting the regions state and split flag updates would sort split failures at different scenarios. It also does not seem consistent the way we do these updates in meta only and don't reflect it on the "in-memory" region info master has.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  22m 10s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -0 :warning: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ branch-1 Compile Tests _ |
| +0 :ok: |  mvndep  |   1m 25s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   7m 34s |  branch-1 passed  |
| +1 :green_heart: |  compile  |   1m  1s |  branch-1 passed with JDK v1.8.0_242  |
| +1 :green_heart: |  compile  |   1m 11s |  branch-1 passed with JDK v1.7.0_252  |
| +1 :green_heart: |  checkstyle  |   2m  5s |  branch-1 passed  |
| +1 :green_heart: |  shadedjars  |   2m 59s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 56s |  branch-1 passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javadoc  |   1m  8s |  branch-1 passed with JDK v1.7.0_252  |
| +0 :ok: |  spotbugs  |   2m 40s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   4m  4s |  branch-1 passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 17s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 58s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  0s |  the patch passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javac  |   1m  0s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  8s |  the patch passed with JDK v1.7.0_252  |
| +1 :green_heart: |  javac  |   1m  8s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   2m  2s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   2m 49s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |   4m 51s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2.  |
| +1 :green_heart: |  javadoc  |   0m 52s |  the patch passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javadoc  |   1m  8s |  the patch passed with JDK v1.7.0_252  |
| +1 :green_heart: |  findbugs  |   4m 23s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   2m 42s |  hbase-client in the patch passed.  |
| -1 :x: |  unit  | 148m 43s |  hbase-server in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 57s |  The patch does not generate ASF License warnings.  |
|  |   | 221m 19s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.client.TestRestoreSnapshotFromClient |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1071/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1071 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux d918b4e4f0ab 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/Base-PreCommit-GitHub-PR_PR-1071/out/precommit/personality/provided.sh |
| git revision | branch-1 / 99a59cf |
| Default Java | 1.7.0_252 |
| Multi-JDK versions | /usr/lib/jvm/zulu-8-amd64:1.8.0_242 /usr/lib/jvm/zulu-7-amd64:1.7.0_252 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1071/3/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1071/3/testReport/ |
| Max. process+thread count | 4229 (vs. ulimit of 10000) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1071/3/console |
| versions | git=1.9.1 maven=3.0.5 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

thangTang commented on hbase-server/src/main/java/org/apache/hadoop/hbase/master/RegionStates.java:
```diff
@@ -758,8 +764,20 @@ public void regionOffline(
           } catch (KeeperException ke) {
             server.abort("Unexpected ZK exception deleting node " + region, ke);
+          } catch (IOException e) {
+            LOG.warn("get daughter from meta exception " + region, e);
```
Yes, but at this time(after createDaughters call completed inside SplitTransactionImpl.execute) the split transaction is still not entirely complete.

wchevreuil replied:
> Yes, but at this time(after createDaughters call completed inside SplitTransactionImpl.execute) the split transaction is still not entirely complete.

And that's why seems wrong to have **SPLIT=true** already at that point.

thangTang replied:
If we update the meta information later, we can only put it after the completion of execute `openDaughters`. It is not very clear to me what impact this might have now. I may need more detailed and comprehensive thinking. My idea is to merge this patch first, at least it can solve most of the problems. Do you think it is okay? @wchevreuil

wchevreuil replied:
> If we update the meta information later, we can only put it after the completion of execute openDaughters.

Yeah, that's what I think. 

> It is not very clear to me what impact this might have now. I may need more detailed and comprehensive thinking. My idea is to merge this patch first, at least it can solve most of the problems. Do you think it is okay?

If you think it's too much work try fixing the state/split flag updates on this PR, then yeah, we can merge this one for now, then work on the other solution in a different jira/RP.

thangTang replied:
> If you think it's too much work try fixing the state/split flag updates on this PR, then yeah, we can merge this one for now, then work on the other solution in a different jira/RP.
Yeah, I hope to merge this patch first, and other work can be completed in new JIRA. At least it can solve the problem of data loss. In extreme scenarios, the region hole can be temporarily repaired by HBCK.

thangTang commented:
> This seems to solve the main issue, which is the parent and its interim daughters been wiped off. However, I'm wondering if the problem happens because we are setting parent state offline and split=true too soon in the split operation. What would you think, @thangTang ?
> 
> Also, is it possible to provide a UT to reproduce? Might be too complex, though, since it involves some sort of race condition.

I think the direct reason for this problem is that the information in the meta table has not been updated after the daughter region is cleaned up. At this time, the sub daughter no longer exists and the master will also try to reassign the parent region, so updating the information in the meta table should not have a negative impact. As for whether to adjust the timing of setting the parent region state in the meta table, we may need to sort out the split process as a whole. After all, when ZK assign is used, split transaction is complex.

I've considered the problem of unit testing, and it's difficult to mock this case --- I met this problem in an extreme scenario --- RS crash when split step after PONR but not entirely complete, no other RS is available, and the master switch occurs.

saintstack commented on hbase-server/src/main/java/org/apache/hadoop/hbase/master/RegionStates.java:
```diff
@@ -737,11 +741,13 @@ public void regionOffline(
     // ConcurrentModificationException and deadlock in case of meta anassigned,
     // but RegionState a blocked.
     Set<HRegionInfo> regionsToOffline = new HashSet<HRegionInfo>();
+    Map<String, HRegionInfo> daughter2Parent = new HashMap<>();
```
Suggest this needs more comment on why we need this accounting in daughter2Parent.

saintstack commented:
Just to be explicit, this issue needs @wchevreuil  approval too before can be merged (I like his comments)

thangTang commented:
@saintstack hi, can you help me to merge this patch? @wchevreuil has approved. After that we will 

> think on correcting the state/split flag updates on a separate jira

wchevreuil commented:
Thanks for the contribution, @thangTang . I had merged this PR into branch-1.

thangTang commented:
> Thanks for the contribution, @thangTang . I had merged this PR into branch-1.

This is my first patch. I'm very happy to have a good start. Thank you and @wchevreuil very much :)
